File: learn-greek-easy-backend/src/schemas/feedback.py
"""Feedback-related Pydantic schemas for API request/response validation.

This module contains schemas for:
- Feedback CRUD operations
- Voting operations
- Paginated feedback listings
"""

from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from src.db.models import FeedbackCategory, FeedbackStatus, VoteType

# ============================================================================
# Author Schema (Brief User Info)
# ============================================================================


class AuthorBriefResponse(BaseModel):
    """Brief author information for feedback items."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    full_name: Optional[str] = None


# ============================================================================
# Feedback Schemas
# ============================================================================


class FeedbackCreate(BaseModel):
    """Schema for creating new feedback."""

    title: str = Field(..., min_length=3, max_length=255)
    description: str = Field(..., min_length=10, max_length=5000)
    category: FeedbackCategory


class FeedbackUpdate(BaseModel):
    """Schema for updating feedback (partial update)."""

    title: Optional[str] = Field(None, min_length=3, max_length=255)
    description: Optional[str] = Field(None, min_length=10, max_length=5000)


class FeedbackResponse(BaseModel):
    """Schema for feedback response."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    description: str
    category: FeedbackCategory
    status: FeedbackStatus
    vote_count: int
    user_vote: Optional[VoteType] = None  # Current user's vote on this item
    admin_response: Optional[str] = None
    admin_response_at: Optional[datetime] = None
    author: AuthorBriefResponse
    created_at: datetime
    updated_at: datetime


class FeedbackListResponse(BaseModel):
    """Schema for paginated feedback list."""

    total: int
    page: int
    page_size: int
    items: List[FeedbackResponse]


# ============================================================================
# Vote Schemas
# ============================================================================


class VoteRequest(BaseModel):
    """Schema for voting on feedback."""

    vote_type: VoteType


class VoteResponse(BaseModel):
    """Schema for vote operation response."""

    feedback_id: UUID
    vote_type: Optional[VoteType]  # None if vote was removed
    new_vote_count: int


# ============================================================================
# Admin Feedback Schemas
# ============================================================================


class AdminFeedbackUpdate(BaseModel):
    """Schema for admin updating feedback (status and/or response)."""

    status: Optional[FeedbackStatus] = None
    admin_response: Optional[str] = Field(None, max_length=500)


class AdminFeedbackResponse(BaseModel):
    """Schema for admin feedback response with admin-specific fields."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    description: str
    category: FeedbackCategory
    status: FeedbackStatus
    vote_count: int
    admin_response: Optional[str] = None
    admin_response_at: Optional[datetime] = None
    author: AuthorBriefResponse
    created_at: datetime
    updated_at: datetime


class AdminFeedbackListResponse(BaseModel):
    """Schema for paginated admin feedback list."""

    total: int
    page: int
    page_size: int
    items: List[AdminFeedbackResponse]

File: learn-greek-easy-frontend/src/types/feedback.ts
// src/types/feedback.ts

/**
 * Feedback Types
 *
 * TypeScript type definitions for the feedback/bug report feature.
 */

// ============================================
// Enums
// ============================================

export type FeedbackCategory = 'feature_request' | 'bug_incorrect_data';

export type FeedbackStatus =
  | 'new'
  | 'under_review'
  | 'planned'
  | 'in_progress'
  | 'completed'
  | 'cancelled';

export type VoteType = 'up' | 'down';

// ============================================
// Response Types
// ============================================

/**
 * Brief author information
 */
export interface AuthorBrief {
  id: string;
  full_name: string | null;
}

/**
 * Single feedback item response
 */
export interface FeedbackItem {
  id: string;
  title: string;
  description: string;
  category: FeedbackCategory;
  status: FeedbackStatus;
  vote_count: number;
  user_vote: VoteType | null; // Current user's vote
  author: AuthorBrief;
  created_at: string;
  updated_at: string;
  admin_response: string | null;
  admin_response_at: string | null;
}

/**
 * Paginated feedback list response
 */
export interface FeedbackListResponse {
  total: number;
  page: number;
  page_size: number;
  items: FeedbackItem[];
}

/**
 * Vote operation response
 */
export interface VoteResponse {
  feedback_id: string;
  vote_type: VoteType | null;
  new_vote_count: number;
}

// ============================================
// Request Types
// ============================================

/**
 * Create feedback request
 */
export interface CreateFeedbackRequest {
  title: string;
  description: string;
  category: FeedbackCategory;
}

/**
 * Update feedback request (partial update)
 */
export interface UpdateFeedbackRequest {
  title?: string;
  description?: string;
}

/**
 * Vote request
 */
export interface VoteRequest {
  vote_type: VoteType;
}

// ============================================
// Filter/Sort Types
// ============================================

export type FeedbackSortField = 'votes' | 'created_at';
export type SortOrder = 'asc' | 'desc';

/**
 * Parameters for listing feedback
 */
export interface FeedbackListParams {
  category?: FeedbackCategory;
  status?: FeedbackStatus;
  sort?: FeedbackSortField;
  order?: SortOrder;
  page?: number;
  page_size?: number;
}

/**
 * Filter state for UI
 */
export interface FeedbackFilters {
  category: FeedbackCategory | null;
  status: FeedbackStatus | null;
  sort: FeedbackSortField;
  order: SortOrder;
}

// ============================================
// UI Helper Types
// ============================================

/**
 * Status display configuration
 */
export interface StatusConfig {
  label: string;
  color: string;
  bgColor: string;
}

/**
 * Status display mapping
 */
export const STATUS_CONFIG: Record<FeedbackStatus, StatusConfig> = {
  new: {
    label: 'New',
    color: 'text-gray-700',
    bgColor: 'bg-gray-100',
  },
  under_review: {
    label: 'Under Review',
    color: 'text-blue-700',
    bgColor: 'bg-blue-100',
  },
  planned: {
    label: 'Planned',
    color: 'text-purple-700',
    bgColor: 'bg-purple-100',
  },
  in_progress: {
    label: 'In Progress',
    color: 'text-yellow-700',
    bgColor: 'bg-yellow-100',
  },
  completed: {
    label: 'Completed',
    color: 'text-green-700',
    bgColor: 'bg-green-100',
  },
  cancelled: {
    label: 'Cancelled',
    color: 'text-red-700',
    bgColor: 'bg-red-100',
  },
};

/**
 * Category display mapping
 */
export const CATEGORY_CONFIG: Record<FeedbackCategory, { label: string; icon: string }> = {
  feature_request: {
    label: 'Feature Request',
    icon: 'lightbulb',
  },
  bug_incorrect_data: {
    label: 'Bug / Incorrect Data',
    icon: 'bug',
  },
};

// ============================================
// Filter Constants for UI
// ============================================

/**
 * Category options for dropdowns/filters
 */
export const FEEDBACK_CATEGORIES = [
  { value: 'feature_request' as const, label: 'Feature Request' },
  { value: 'bug_incorrect_data' as const, label: 'Bug / Incorrect Data' },
] as const;

/**
 * Status options for dropdowns/filters
 */
export const FEEDBACK_STATUSES = [
  { value: 'new' as const, label: 'New' },
  { value: 'under_review' as const, label: 'Under Review' },
  { value: 'planned' as const, label: 'Planned' },
  { value: 'in_progress' as const, label: 'In Progress' },
  { value: 'completed' as const, label: 'Completed' },
  { value: 'cancelled' as const, label: 'Cancelled' },
] as const;

// ============================================
// Admin Feedback Types
// ============================================

/**
 * Admin feedback item (without user_vote since admin doesn't vote)
 */
export interface AdminFeedbackItem {
  id: string;
  title: string;
  description: string;
  category: FeedbackCategory;
  status: FeedbackStatus;
  vote_count: number;
  author: AuthorBrief;
  created_at: string;
  updated_at: string;
  admin_response: string | null;
  admin_response_at: string | null;
}

/**
 * Paginated admin feedback list response
 */
export interface AdminFeedbackListResponse {
  total: number;
  page: number;
  page_size: number;
  items: AdminFeedbackItem[];
}

/**
 * Admin feedback update request
 */
export interface AdminFeedbackUpdateRequest {
  status?: FeedbackStatus;
  admin_response?: string;
}

/**
 * Parameters for admin feedback listing
 */
export interface AdminFeedbackListParams {
  status?: FeedbackStatus;
  category?: FeedbackCategory;
  page?: number;
  page_size?: number;
}
